rfcs/20201011-max-pooling-with-global-indices/README.md

```diff
@@ -0,0 +1,274 @@
+        const dnnl_memory_desc_t *indices_desc, const dnnl_dims_t strides,
+        const dnnl_dims_t kernel, const dnnl_dims_t dilation,
+        const dnnl_dims_t padding_l, const dnnl_dims_t padding_r);
+~~~
+
+### Option 2 - new primitive without a new alg_kind.
+
+This option coincides with Option 1, but makes `dnnl_pooling_max` algorithm to
+dump global indices unconditionally instead, requiring users to modify the code
+and specify the second input, which will break existing pooling API and is not
+user-friendly. Another drawback is sacrificing performance for existing
+implementation.
+
+### Option 3 - new algorithm and attributes extension only.
+
+The alternative of putting `indices_desc` into a new pooling descriptor to
+represent a memory layout through attributes as well. This would look like
+binary post-ops API. Since we need only memory format and data type, this could
+utilize `dnnl_format_tag_t tag` value and/or `dnnl_dims_t strides` array and
+`dnnl_data_type_t data_type` value instead a whole memory descriptor.
+
+No API example provided yet as requires broad group discussion whether this
+option is viable.
+
+## Implementation
+
+Disclaimer: this paragraph does not belong to the proposal itself but sheds some
+light on potential implementation choices just for information purposes.
+
+The idea is to keep global indices internally for spatial case and when the
+outer parallel loops finished, modify the value if `c_spatial` or `all_dims`
+options was requested. It also will reorder the memory layout if it was asked
+to be different from destination memory layout, since using same formats
+internally will help to speed up computations.
+
+## Backward
+
+Existing max pooling implementation for best performance purpose requires
+passing a workspace, which signature is encrypted from an end user and based on
+the implementation. The idea is still to request the workspace on forward
+despite the fact the global indices were requested. This demand allows the user
+to use faster implementation on backward by using `dnnl_pooling_max` algorithm
+and passing the workspace (not global indices) since there is an expectation
+that the library will still have a single JIT-ted implementation for all
+pooling flavors.
+
+In case `dnnl_pooling_max_with_indices` is requested on backward, it will demand
+for global indices memory and provide an implementation using global indices
+(not workspace), which is expected to perform worse (though it is a pure
+speculation at this point of time).
+
+## Testing
+
+Benchdnn will require attributes extension, pooling driver will require new
+options to pass the flavor of global indices output and the order. Reference
+implementation will require adjustment to respect those new options and their
+values as well. Further comparison will consider point-to-point comparison of
+both destination value and global index value.
+
+## Open Question
+
+* Are getter functions required for attributes extension?
```

Comment:
I would vote for adding getter functions for API completeness and simplifying interface tests of set functions. Also it is pretty straight forward from implementation point of view.

Reply:
Alright, updated an RFC with getter counterparts, thanks.